[JSC] Optimize `Array#at` in DFG/FTL

sosukesuzuki opened the pull request:
#### 8db8cb832a122e2873b961ecae67c5084155c43d
<pre>
[JSC] Optimize `Array#at` in DFG/FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=285095">https://bugs.webkit.org/show_bug.cgi?id=285095</a>

Reviewed by NOBODY (OOPS!).

This patch optimize `Array#at` in DFG/FTL using same approach as
`GetByVal` and `StringAt`.

                                                  TipOfTree                  Patched

array-prototype-at-out-of-bounds-contiguous
                                                1.2855+-0.0333     ^      0.9349+-0.0073        ^ definitely 1.3750x faster
array-prototype-at-in-bounds-contiguous         1.4921+-0.0096     ^      0.9417+-0.0087        ^ definitely 1.5844x faster
array-prototype-at-in-bounds-contiguous-neg
                                                1.7482+-0.0216     ^      0.9724+-0.0266        ^ definitely 1.7979x faster
array-prototype-at-out-of-bounds-contiguous-neg
                                                1.3912+-0.0463     ^      0.9658+-0.0137        ^ definitely 1.4404x faster
array-prototype-at-in-bounds-double-neg         1.7896+-0.1390     ^      0.9218+-0.0887        ^ definitely 1.9415x faster
array-prototype-at-out-of-bounds-double-neg
                                                1.3934+-0.0361     ^      0.9822+-0.0199        ^ definitely 1.4187x faster
array-prototype-at-in-bounds-int32              1.5240+-0.0864     ^      0.9183+-0.0177        ^ definitely 1.6597x faster
array-prototype-at-in-bounds-int32-neg          1.7369+-0.0131     ^      0.9576+-0.0080        ^ definitely 1.8138x faster
array-prototype-at-out-of-bounds-int32          1.2567+-0.0136     ^      0.9449+-0.0090        ^ definitely 1.3299x faster
array-prototype-at-out-of-bounds-int32-neg
                                                1.3925+-0.0366     ^      0.9708+-0.0226        ^ definitely 1.4344x faster
array-prototype-at-out-of-bounds-double         1.3008+-0.1358     ^      0.9612+-0.0058        ^ definitely 1.3533x faster
array-prototype-at-in-bounds-double             1.5088+-0.0320     ^      0.9846+-0.0260        ^ definitely 1.5325x faster

* JSTests/microbenchmarks/array-prototype-at-in-bounds-contiguous-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-contiguous.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-double-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-double.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-int32-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-int32.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-contiguous-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-contiguous.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-double-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-double.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-int32-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-int32.js: Added.
(test):
* JSTests/stress/array-prototype-at-out-of-bounds-contiguous.js: Added.
(shouldBe):
(test):
(i.shouldBe.test):
* JSTests/stress/array-prototype-at-out-of-bounds-double.js: Added.
(shouldBe):
(test):
* JSTests/stress/array-prototype-at-out-of-bounds-int32.js: Added.
(shouldBe):
(test):
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(at): Deleted.
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
(JSC::DFG::BackwardsPropagationPhase::propagate):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasHeapPrediction):
(JSC::DFG::Node::hasStorageChild const):
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValImpl):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayAt):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayPush):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/Intrinsic.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8db8cb832a122e2873b961ecae67c5084155c43d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84978 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4707 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39374 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90127 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36034 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4798 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12683 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66131 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23949 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88023 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3669 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77228 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46397 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3549 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31456 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35107 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77946 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74352 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32266 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91500 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/84023 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12322 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9006 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74616 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12551 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73039 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73736 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18125 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16577 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4345 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12270 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17723 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106413 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12106 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25686 "Found 11 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/function-apply-many-args.js.layout-no-llint, stress/json-stringify-inspector-check.js.no-llint, stress/regexp-escape-oom.js.default, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm, wasm.yaml/wasm/stress/wasm-bbq-catch-unbind.js.default-wasm, wasm.yaml/wasm/stress/wasm-bbq-catch-unbind.js.wasm-bbq, wasm.yaml/wasm/stress/wasm-bbq-catch-unbind.js.wasm-collect-continuously, wasm.yaml/wasm/stress/wasm-bbq-catch-unbind.js.wasm-eager, wasm.yaml/wasm/stress/wasm-bbq-catch-unbind.js.wasm-eager-jettison, wasm.yaml/wasm/stress/wasm-bbq-catch-unbind.js.wasm-no-cjit ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15602 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13851 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGNodeType.h:
```diff
@@ -255,6 +255,7 @@ namespace JSC { namespace DFG {
     macro(CheckArrayOrEmpty, NodeMustGenerate) \
     macro(GetPrivateName, NodeResultJS | NodeMustGenerate) \
     macro(GetPrivateNameById, NodeResultJS | NodeMustGenerate) \
+    macro(ArrayAt, NodeResultJS | NodeHasVarArgs) \
```
NodeMustGenerate is necessary

sosukesuzuki replied:
Fixed

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -2742,9 +2745,11 @@ void SpeculativeJIT::compileGetByVal(Node* node, const ScopedLambda<std::tuple<J
         if (node->arrayMode().isInBounds()) {
             SpeculateStrictInt32Operand property(this, m_graph.varArgChild(node, 1));
             StorageOperand storage(this, m_graph.varArgChild(node, 2));
+            GPRTemporary propertyTemp(this);
```
We should not add new register allocation for GetByVal etc. as it regress code quality for the one of the hottest code. These things must be done only for ArrayAt.

sosukesuzuki replied:
I've created a new function `compileArrayAt`

sosukesuzuki commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -2700,15 +2921,10 @@ void SpeculativeJIT::compileGetByVal(Node* node, const ScopedLambda<std::tuple<J
-                BaselineJITRegisters::GetByVal::baseJSR.payloadGPR(),
-                BaselineJITRegisters::GetByVal::propertyJSR.payloadGPR(),
-            });
+        shuffleRegisters<GPRReg, 2>({ baseGPR, propertyGPR }, {
+            BaselineJITRegisters::GetByVal::baseJSR.payloadGPR(),
+            BaselineJITRegisters::GetByVal::propertyJSR.payloadGPR(),
+        });
```
via `check-webkit-style`

kmiller68 commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -2821,6 +2821,44 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
+        case Array::Int32:
+            if (arrayMode.isEffectfulOutOfBounds()) {
+                clobberWorld();
+                makeHeapTopForNode(node);
```
Nit: It might be slightly less code to hoist this out of the switch.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -2578,6 +2578,56 @@ auto ByteCodeParser::handleIntrinsicCall(Node* callee, Operand resultOperand, Ca
+            case Array::Int32:
+            case Array::Contiguous: {
+                JSGlobalObject* globalObject = m_graph.globalObjectFor(currentNodeOrigin().semantic);
+                if (globalObject->arrayPrototypeChainIsSaneWatchpointSet().state() == IsWatched) {
```
I think we might as well check if we're having a bad time too.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGClobberize.h:
```diff
@@ -990,6 +991,58 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
         return;
     }
+
+    case ArrayAt: {
```
Is this different than the `GetByVal` case below? I think we could just assert it's one of the modes you have here and `FALLTHROUGH`

kmiller68 commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -2578,6 +2578,56 @@ auto ByteCodeParser::handleIntrinsicCall(Node* callee, Operand resultOperand, Ca
+            if (argumentCountIncludingThis < 2)
+                return CallOptimizationResult::DidNothing;
+
+            if (m_inlineStackTop->m_exitProfile.hasExitSite(m_currentIndex, BadType))
```
I wonder if it's worth either profiling or just speculating that the index is non-negative. I imagine a lot of `Array.prototype.at` users are just using it without the reverse indexing feature. Can you file a FIXME for that and comment here?

sosukesuzuki replied:
>  I imagine a lot of Array.prototype.at users are just using it without the reverse indexing feature.

Hmm, I imagine that users who use `Array.prototype.at` instead of normal index access will want to use negative indexes.

kmiller68 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -5943,19 +5946,33 @@ IGNORE_CLANG_WARNINGS_END
             if (arrayMode.isInBounds()) {
-                LValue result = m_out.load64(baseIndexWithProvenValue(heap, storage, index, m_graph.varArgChild(m_node, 1)));
+                if (m_node->op() == ArrayAt)
+                    speculate(OutOfBounds, noValue(), nullptr, m_out.aboveOrEqual(index, arrayLength));
```
Doesn't CheckInBounds force you to OSR exit here when the index is negative? If I recall correctly, CheckInBounds always exits on negative indices. I think you need to teach check in bounds about `at`s case or just assume the index is non-negative for the in bounds case. If you do the latter I think you just don't need this check, although it might be worth it as a debug assert.

sosukesuzuki replied:
My current patch does not use the `CheckInBounds` node. I didn't know that.

Also, in my opinion, we should not assume that the index is non-negative (https://github.com/WebKit/WebKit/pull/38331#discussion_r1917851552).

Ideally, I think we should introduce a new `CheckInBoundsNegative` node and insert it in `DFGSSALoweringPhase`. This could also be used for the FTL for `String#At` that was introduced in https://commits.webkit.org/287653@main.

However, adding a new node would involve a amount of changes. So, I’d prefer to leave this patch as is, mark it with a `FIXME`( https://bugs.webkit.org/show_bug.cgi?id=286056 ), and handle this in a separate patch.

As far as I understand, this patch should work as expected as is.

kmiller68 replied:
IIRC, one of the reasons to use `at` is that it works with containers that don't support array access. But upon further searching those don't seem to exist. In that case, it's not with adding an optimization until we know folks use `at` in that way.

Ah I was mistaken, I thought CheckInBounds was automatically inserted for arrayMode nodes in the FTL. Seems like that's not the case. A FIXME is fine. Although I think I'd implement `at` with a `Select` node rather than a separate `CheckInBoundsNegative`. Which, thinking about it more would mean you don't have to have a `ArrayAt` node at all. So maybe that's the right approach for this whole PR...

sosukesuzuki replied:
>  Although I think I'd implement at with a Select node rather than a separate CheckInBoundsNegative. Which, thinking about it more would mean you don't have to have a ArrayAt node at all. So maybe that's the right approach for this whole PR...

Thank you for the suggestion! This is my first time hearing about the `Select` node. Could you explain a bit more about how it works and how it could be applied here?

kmiller68 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -6004,14 +6021,22 @@ IGNORE_CLANG_WARNINGS_END
 
+            LValue arrayLength = m_out.load32NonNegative(storage, m_heaps.Butterfly_publicLength);
+            LValue index = m_node->op() == ArrayAt ?
+                m_out.select(m_out.lessThan(originalIndex, m_out.int32Zero), m_out.add(arrayLength, originalIndex), originalIndex) : originalIndex;
```
Nit: I would just put this on one line

kmiller68 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -5943,19 +5946,33 @@ IGNORE_CLANG_WARNINGS_END
+                LValue result = m_out.load64(
+                    m_node->op() == ArrayAt
+                        ? m_out.baseIndex(heap, storage, m_out.zeroExtPtr(index))
+                        : baseIndexWithProvenValue(heap, storage, index, m_graph.varArgChild(m_node, 1)));
```
I think in this case we know we're using originalIndex so we can just use `baseIndexWithProvenValue(heap, storage, originalIndex, m_graph.varArgChild(m_node, 1)));`

kmiller68 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -6004,14 +6021,22 @@ IGNORE_CLANG_WARNINGS_END
 
             if (arrayMode.isInBounds()) {
+                if (m_node->op() == ArrayAt)
+                    speculate(OutOfBounds, noValue(), nullptr, m_out.aboveOrEqual(index, arrayLength));
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -6004,14 +6021,22 @@ IGNORE_CLANG_WARNINGS_END
-                    baseIndexWithProvenValue(heap, storage, index, m_graph.varArgChild(m_node, 1)));
+                    m_node->op() == ArrayAt
+                        ? m_out.baseIndex(heap, storage, m_out.zeroExtPtr(index))
+                        : baseIndexWithProvenValue(heap, storage, index, m_graph.varArgChild(m_node, 1)));
```
Ditto.

sosukesuzuki committed:
[JSC] Optimize `Array#at` in DFG/FTL
https://bugs.webkit.org/show_bug.cgi?id=285095

Reviewed by NOBODY (OOPS!).

This patch optimize `Array#at` in DFG/FTL using same approach as
`GetByVal` and `StringAt`.

                                                  TipOfTree                  Patched

array-prototype-at-out-of-bounds-contiguous
                                                1.2855+-0.0333     ^      0.9349+-0.0073        ^ definitely 1.3750x faster
array-prototype-at-in-bounds-contiguous         1.4921+-0.0096     ^      0.9417+-0.0087        ^ definitely 1.5844x faster
array-prototype-at-in-bounds-contiguous-neg
                                                1.7482+-0.0216     ^      0.9724+-0.0266        ^ definitely 1.7979x faster
array-prototype-at-out-of-bounds-contiguous-neg
                                                1.3912+-0.0463     ^      0.9658+-0.0137        ^ definitely 1.4404x faster
array-prototype-at-in-bounds-double-neg         1.7896+-0.1390     ^      0.9218+-0.0887        ^ definitely 1.9415x faster
array-prototype-at-out-of-bounds-double-neg
                                                1.3934+-0.0361     ^      0.9822+-0.0199        ^ definitely 1.4187x faster
array-prototype-at-in-bounds-int32              1.5240+-0.0864     ^      0.9183+-0.0177        ^ definitely 1.6597x faster
array-prototype-at-in-bounds-int32-neg          1.7369+-0.0131     ^      0.9576+-0.0080        ^ definitely 1.8138x faster
array-prototype-at-out-of-bounds-int32          1.2567+-0.0136     ^      0.9449+-0.0090        ^ definitely 1.3299x faster
array-prototype-at-out-of-bounds-int32-neg
                                                1.3925+-0.0366     ^      0.9708+-0.0226        ^ definitely 1.4344x faster
array-prototype-at-out-of-bounds-double         1.3008+-0.1358     ^      0.9612+-0.0058        ^ definitely 1.3533x faster
array-prototype-at-in-bounds-double             1.5088+-0.0320     ^      0.9846+-0.0260        ^ definitely 1.5325x faster

* JSTests/microbenchmarks/array-prototype-at-in-bounds-contiguous-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-contiguous.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-double-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-double.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-int32-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-in-bounds-int32.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-contiguous-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-contiguous.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-double-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-double.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-int32-neg.js: Added.
(test):
* JSTests/microbenchmarks/array-prototype-at-out-of-bounds-int32.js: Added.
(test):
* JSTests/stress/array-prototype-at-out-of-bounds-contiguous.js: Added.
(shouldBe):
(test):
(i.shouldBe.test):
* JSTests/stress/array-prototype-at-out-of-bounds-double.js: Added.
(shouldBe):
(test):
* JSTests/stress/array-prototype-at-out-of-bounds-int32.js: Added.
(shouldBe):
(test):
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(at): Deleted.
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
(JSC::DFG::BackwardsPropagationPhase::propagate):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasHeapPrediction):
(JSC::DFG::Node::hasStorageChild const):
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValImpl):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayAt):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayPush):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/Intrinsic.h:
